ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,177 @@ func (ps *prioritySampler) apply(spn *span) {
 	}
 	spn.SetTag(keySamplingPriorityRate, rate)
 }
+
+type rulesSampler struct {
+	rules   []samplingRule
+	rate    float64
+	limiter *rate.Limiter
+	// "effective rate" calculations
+	mu           sync.Mutex
+	ts           int64
+	allowed      int
+	total        int
+	previousRate float64
+}
+
+func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	rate := sampleRate()
+	return &rulesSampler{
+		rules:   samplingRules(rules),
+		rate:    rate,
+		limiter: rateLimiter(rate),
+	}
+}
+
+// samplingRules validates the user-provided rules and returns an internal representation.
+// If the DD_TRACE_SAMPLING_RULES environment variable is set, then the rules from
+// tracer.WithSamplingRules are ignored.
+func samplingRules(rules []SamplingRule) []samplingRule {
```

Comment:
Would makeSamplingRules or newSamplingRules be better?

Reply:
Or validSamplingRules or activeSamplingRules. It does a bit more than just making the value.

Reply:
Right, then `newSamplingRules` should be good. Also, should we have `[]*samplingRule`? This will be iterated over on every span and it will be copied each time if we don't use pointers.